### PR TITLE
1st order cky implementation

### PR DESCRIPTION
**Pull request by teffland**
Hi,

I'd like to contribute this implementation of a first-order cky-style crf with anchored rule potentials: $\phi[i,j,k,A,B,C] := \phi(A_{i,j} \rightarrow B_{i,k}, C{k+1,j})$.

I also added code to the `_Struct` class that allows calculating marginals even if input tensors don't require a gradient (i.e., after `model.eval()`)

Please let me know if you'd like to see any changes.

Thanks,
Tom

### CODE REVIEW

**Review by srush — CHANGES_REQUESTED** [review]
This looks great! Very excited to add. 

Couple questions: 

* Reshape always worries me. Can those be views? 
* Repeat is almost never necessary. What operators is failing there? 
* I would prefer the sr.sum be replace with a view and then a single sum. 

Were you able to run any tests? I know travis is broken (I need to move to github), but it would be great to have at least one test. For other algorithms I enumerate all examples and score. Can you try that?

**Comment by teffland** [comment]
Great, excited to contribute!

Re: your questions

* Reshape: I can definitely try this with view. Just curious, why does reshape worry you? I thought reshape was just a more robust version of view that creates a contiguous copy in cases where view would fail. 
* Repeat: true, I can try without and it should work I think, I guess I just naturally tend towards implementations that go through all of the steps explicitly but I should get better about trusting broadcasting.
* One sum: this is definitely doable, will make that change
* I did write some tests locally in a notebook a while back. I can dig those up and incorporate them. By enumerate and score, you mean check that the dp partition matches the partition computed by explicitly enumerating all structures, correct?

Thanks for the feedback,
Tom

**Comment by srush** [comment]
I can help with testing. Yes, all the algorithms in torch struct test that the enumeration has the same marginals. Its helped catch a ton of bugs. 

* Reshape- maybe they changed this, earlier version caused problems.

**Comment by teffland** [comment]
Cool, help with the testing would nice but I can also take a crack at it later. When I get to it I'll also incorporate the suggested changes, (and will try it out with view).

**Comment by srush** [comment]
Hey sorry, I flaked on helping out here. Send me an email if you want to work on it more. I'm going to do some repo maintenance this week.

**Comment by teffland** [comment]
Hey, no worries same here. Will send an email

**Comment by teffland** [comment]
included in https://github.com/harvardnlp/pytorch-struct/pull/93

**Comment by srush** [comment]
(still not merged, keeping open)
